util/contextsleep.go

```diff
@@ -0,0 +1,16 @@
+package util
+
+import (
+	"context"
+	"time"
+)
+
+// ContextSleep will sleep for the specified duration or until the context is canceled.
+func ContextSleep(ctx context.Context, d time.Duration) error {
```

Comment:
**question:** this is a nice way to cleanly handle both of: checking for context cancel and delaying the next loop execution. I do have slight concerns that the intent of the function is not totally clear. Are there other contexts that we would use this other than a retry or wait loop?

Reply:
It's basically a context-aware version of `time.Sleep`. When a request is cancelled, connection failed, or app is shutting down; time.Sleep will continue to wait. This just allows for an early exit as well as a way to break the loop if the context is cancelled.

There are actually a number of loops in GoAlert that implement this, or something very similar. But the only one I can think of that isn't a retry/wait loop is in Basic auth where we implement a random delay for failed password attempts that we'd want to immediately abort when the request is cancelled.

Reply:
Sounds great, hopefully we can start reusing this in some of those places too. It's nice that this can just use `time.After` now too, cleaning up timers was a pain.